src/pointer/ptr.rs

```diff
@@ -0,0 +1,1010 @@
+    // for unique pointers, since at most one may exist at a time.
+    impl<
+            'a,
+            T,
+            const ALIGNMENT_INVARIANT: invariant::Alignment,
+            const VALIDITY_INVARIANT: invariant::Validity,
+        > Clone for Ptr<'a, T, { invariant::Shared }, ALIGNMENT_INVARIANT, VALIDITY_INVARIANT>
+    where
+        T: 'a + ?Sized,
+    {
+        #[inline]
+        fn clone(&self) -> Self {
+            *self
+        }
+    }
+
+    impl<
+            'a,
+            T,
+            const ALIASING_INVARIANT: invariant::Aliasing,
+            const ALIGNMENT_INVARIANT: invariant::Alignment,
+            const VALIDITY_INVARIANT: invariant::Validity,
+        > Debug for Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, VALIDITY_INVARIANT>
+    where
+        T: 'a + ?Sized,
+    {
+        #[inline]
+        fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+            self.as_non_null().fmt(f)
+        }
+    }
+}
+
+/// Methods for converting to and from `Ptr` and Rust's safe reference types.
+mod _conversions {
+    use super::*;
+
+    /// `&'a T` → `Ptr<'a, T>`
+    impl<'a, T> Ptr<'a, T, { invariant::Shared }, { invariant::Aligned }, { invariant::Valid }>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Constructs a `Ptr` from a shared reference.
+        #[doc(hidden)]
+        #[inline]
+        pub fn from_ref(ptr: &'a T) -> Self {
+            let ptr = core::ptr::NonNull::from(ptr);
+            // SAFETY:
+            // 0. `ptr`, by invariant on `&'a T`, is derived from some valid
+            //    Rust allocation, `A`.
+            // 1. `ptr`, by invariant on `&'a T`, has valid provenance for `A`.
+            // 2. `ptr`, by invariant on `&'a T`, addresses a byte range which
+            //    is entirely contained in `A`.
+            // 3. `ptr`, by invariant on `&'a T`, addresses a byte range whose
+            //    length fits in an `isize`.
+            // 4. `ptr`, by invariant on `&'a T`, addresses a byte range which
+            //     does not wrap around the address space.
+            // 5. `A`, by invariant on `&'a T`, is guaranteed to live for at
+            //    least `'a`.
+            // 6. `ptr`, by invariant on `&'a T`, conforms to the aliasing
+            //    invariant of `invariant::Shared`.
+            // 7. `ptr`, by invariant on `&'a T`, conforms to the alignment
+            //    invariant of `invariant::Aligned`.
+            // 8. `ptr`, by invariant on `&'a T`, conforms to the validity
+            //    invariant of `invariant::Valid`.
+            unsafe { Self::new(ptr) }
+        }
+    }
+
+    /// `Ptr<'a, T>` → `&'a T`
+    impl<'a, T> Ptr<'a, T, { invariant::Shared }, { invariant::Aligned }, { invariant::Valid }>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Converts the `Ptr` to a shared reference.
+        // This consumes `self`, not `&self`, because `self` is, logically, a
+        // pointer.
```

Comment:
```suggestion
        // This consumes `self`, not `&self`, because `self` is, logically, a
        // pointer. Since this method is only available for `invariant::Shared`,
        // `Self: Copy`, and so this doesn't prevent the caller from still
        // using the pointer after calling `as_ref`.
```